### PR TITLE
Adds Doxygen documentation generation to the codebase

### PR DESCRIPTION
**Pull request by KipHamiltons**
This should have a follow-up PR in NUbook explaining how the system works. I'll write that once we have agreed on an approach/details for this PR.

If anyone has any suggestions for config options to add, let me know. I have also added a couple of TODOs in the doxyfile to chase up.

Further follow ups should work out the following (at least):

- Styling/customised layout and stuff like that
- Protobuf support. Currently doxygen ignores `.proto` files, because protobuf is not a supported language. [this page](https://opensimulationinterface.github.io/osi-documentation/proto2cpp/README.html) indicates that there should be ways. This may warrant looking at sphinx or something though, depending on how hard it is to get protobuf in there
- Integration into nubook (monorepo seems to make things easier in this respect. Netlify builds will need to be sorted out (we do too many builds on this repo for our free tier of builds on netlify. [Netlify github actions](https://github.com/netlify/actions) are definitely worth following up, once #353 is merged/sorted out).
- Actually documenting the code lol. There are very few parts with comments in there. Lots and lots to do...

**EDIT: REVIEWERS NOTE: To test this, it's probably best to just get doxygen on your local machine - unless you want to build a docker image on your machine** :weary: 
To view the generated stuff, open `doc/generated/html/index.html` in your browser.

### CODE REVIEW

**Comment by KipHamiltons** [comment]
Looking at the stuff for the `Webots` class as an example - none of the member variables (all of them have `@brief`s) are showing up. This requires further investigation

**Comment by KipHamiltons** [comment]
I changed a bunch of config, and it's actually pretty nice for the `Webots` class now! Amazing

**Review by ysims — COMMENTED** [review]
Yeah, looks good. Can we get the generated stuff put into the gitignore?

File: .Doxyfile [hunk 1]
```diff
@@ -0,0 +1,2553 @@
...
+                         *.hxx \
+                         *.hpp \
+                         *.h++
+                        # TODO: Fix protobuf
+                        # Look here: https://opensimulationinterface.github.io/osi-documentation/proto2cpp/README.html
+                        #  *.proto
+
+                        # TODO: Investigate these. Are they worth parsing with doxygen?
+                        #  *.md \
+                        #  *.py
```
[the remark this answers is not in the document]
Comment (KipHamiltons) [reply .Doxyfile:893]: I'll make an issue about integrating our in-repo markdown docs once this PR gets merged

**Review by CMurtagh-LGTM — APPROVED** [review]
LGTM

**Pull request merged**